### PR TITLE
Add event to alter delete website confirmation

### PR DESCRIPTION
**Pull request by AltamashShaikh**
### Description:

When a website is about to be deleted Matomo currently shows a confirmation.
As users might not be aware which additional plugin data might be removed with that, plugins should have the possibility to add some additional information/warning to that confirmation.

This PR introduces a new event `SitesManager.getMessagesToWarnOnSiteRemoval`, which allows plugins to provide additional text blocks that should be displayed within the modal.

Fixes: #PG-3574

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: plugins/SitesManager/API.php
Comment (sgiehl) [inline plugins/SitesManager/API.php]: It's imho a bad choice to update/change the API only to inject something that will only be visible in the UI.
For that it's better to introduce a new UI / JavaScript event another plugin can listen to.
Is there are reason why you didn't use `Matomo.postEvent()` directly in the `SiteFields` component?

Comment (AltamashShaikh) [reply plugins/SitesManager/API.php]: @sgiehl Changed to fetch the content via an API call only when delete action is triggered

**Review by sgiehl — CHANGES_REQUESTED** [review]
Hey @AltamashShaikh,
I've discussed the solution again in core team. As mentioned before the "perfect" solution would be using a javascript event, that plugins like the TagManager could easily use to add content to the confirmation. But as we currently do not have something like a global javascript event registry it is not really possible to solve it that way.
Such a javascript event registry would be beneficial for a lot other use cases as well, but core team would need to plan and work on that separately.

To be able to allow altering the removal confirmation already we might indeed need to go with a similar workaround like you implemented. But from our point of view the implementation - especially in TagManager - is too unflexible.
You are currently fully replacing the text of the confirmation. That for sure works, but if another plugin would aim to do the same the result is completely unpredictable. The plugin, that would be loaded last, would most likely "win the race" and place it's content, while the content of other plugins is simply overwritten.
When introducing new events you should always consider that other plugins might want to use those events as well.
My approach here would have been to have an event that only provides sort of a list of things to warn that the will be removed as well and keep a more general basic message as introduction.
So there would be an event like `getThingsToWarnOnSiteRemoval`, which receives the `idSite` and a variable, prefilled with an empty array. TagManager can then add the items that will be removed to that array. If the list contains items after calling the event, core would adjust the message to add them accordingly.

**Comment by AltamashShaikh** [comment]
> roducing new events you should always consider that other plugins might want to use those events as well.
> My approach here would have been to have an event that only provides sort of a list of things to warn that the will be removed as well and keep a more general basic message as introductio

@sgiehl I changed the event to pass an array with a default message that as the first key and plugins can add into the array of update the first key as required, will that work for you guys ??

File: plugins/SitesManager/API.php
Comment (sgiehl) [inline plugins/SitesManager/API.php]:
```suggestion
     * @return string messages to warn users
     * @internal
     * @unsanitized
```

Such API methods should be marked as internal, so they are not listed in API documentation.
Also for new API methods we should use the `@unsanitized` hint, where possible. This will cause parameters provided for the API not to be sent through `Common::sanitizeInputValue`. As we only expect an int value here, there is no risk at all for using the parameters unsanitized, as integers can't contain manipulated content.

File: plugins/SitesManager/API.php
Comment (sgiehl) [inline plugins/SitesManager/API.php]: The initial message should not be part of that array, but should be hard coded in the vue code. Only additional messages should be provided here.

File: plugins/SitesManager/API.php
Comment (sgiehl) [inline plugins/SitesManager/API.php]: Removing a site should only be allowed to the superuser. So we should check for super user instead of view access here.

File: plugins/SitesManager/API.php
Comment (sgiehl) [inline plugins/SitesManager/API.php]: Please try to use type hints where possible

```suggestion
    public function getMessagesToWarnOnSiteRemoval(int $idSite): array
```

File: plugins/SitesManager/API.php
Comment (sgiehl) [inline plugins/SitesManager/API.php]: We should return the array here instead, and let the consumer decide how to put that together.

```suggestion
        return $messages;
```

File: plugins/SitesManager/vue/src/SiteFields/SiteFields.vue
Comment (sgiehl) [inline plugins/SitesManager/vue/src/SiteFields/SiteFields.vue]: according to the suggested changes above this would then need to be:

```suggestion
          this.deleteSiteExplanation += response.value.join('<br>');
```

File: tests/UI/expected-screenshots/UIIntegrationTest_api_listing.png
Comment (sgiehl) [inline tests/UI/expected-screenshots/UIIntegrationTest_api_listing.png:1]: This can be reverted I guess. Once the API is marked as internal it shouldn't be shown here anymore.

**Review by sgiehl — CHANGES_REQUESTED** [review]
I've left a couple of comments for further improvements. Let me know if you have questions on anything...

File: plugins/SitesManager/API.php
Comment (AltamashShaikh) [reply plugins/SitesManager/API.php]: Updated :+1:

File: plugins/SitesManager/API.php
Comment (AltamashShaikh) [reply plugins/SitesManager/API.php]: Removed

File: plugins/SitesManager/API.php
Comment (AltamashShaikh) [reply plugins/SitesManager/API.php]: Updated :+1:

File: tests/UI/expected-screenshots/UIIntegrationTest_api_listing.png
Comment (AltamashShaikh) [reply tests/UI/expected-screenshots/UIIntegrationTest_api_listing.png:1]: Its removed now

**Comment by AltamashShaikh** [comment]
@sgiehl I have updated the code as per your suggestion if all okay please don't merge as I need to get the text checked from product once to ensure if we need to change anything in core.

**Comment by AltamashShaikh** [comment]
@sgiehl All good from product team to merge this :+1: 
Please review and merge this

File: plugins/SitesManager/API.php
Comment (sgiehl) [inline plugins/SitesManager/API.php]: We should add a comment above the event to document what it is good for. Like it'S e.g. the case here:
https://github.com/matomo-org/matomo/blob/9a3ef94df61682e7d543f3f8b438c368f6eb71a6/core/Scheduler/Scheduler.php#L157-L166

And when looking at other events it seems we use to pass the referenced parameter first, and then the other. So you may switch `idSite` and `messages`

Comment (AltamashShaikh) [reply plugins/SitesManager/API.php]: @sgiehl Added as suggested :+1:

**Review by sgiehl — APPROVED** [review]

**Pull request merged**
